pkg/server/datastore/sqlstore/sqlstore.go

```diff
@@ -1757,6 +1766,54 @@ func pruneAttestedNodeEvents(tx *gorm.DB, olderThan time.Duration) error {
 	return nil
 }
 
+func notBanned(tx *gorm.DB) *gorm.DB {
+	return tx.Where("serial_number <> ''")
+}
+
+func expiredForDuration(expiredBefore time.Time) func(db *gorm.DB) *gorm.DB {
+	return func(tx *gorm.DB) *gorm.DB {
+		return tx.Where("expires_at < ?", expiredBefore)
+	}
+}
+
+func includeNonReattestable(include bool) func(db *gorm.DB) *gorm.DB {
+	return func(tx *gorm.DB) *gorm.DB {
+		if !include {
+			return tx.Where("can_reattest = ?", true)
+		} else {
+			return tx
+		}
+	}
+}
+
+func pruneAttestedExpiredNodes(tx *gorm.DB, expiredBefore time.Time, include bool, logger logrus.FieldLogger) error {
+	var expiredNodes []AttestedNode
+
+	if err := tx.Scopes(expiredForDuration(expiredBefore), includeNonReattestable(include), notBanned).Limit(1000).Find(&expiredNodes).Error; err != nil {
+		return newWrappedSQLError(err)
+	}
+
+	for _, node := range expiredNodes {
+		_, err := deleteAttestedNodeAndSelectors(tx, node.SpiffeID)
+		if err != nil {
+			return err
+		}
+		if err := createAttestedNodeEvent(tx, &datastore.AttestedNodeEvent{
+			SpiffeID: node.SpiffeID,
+		}); err != nil {
+			return err
+		}
+
+		logger.WithFields(logrus.Fields{
+			telemetry.SPIFFEID:         node.SpiffeID,
+			telemetry.SVIDSerialNumber: node.SerialNumber,
+			telemetry.SerialNumber:     "CURRENT",
```

Comment:
why is this required?

Reply:
I took pruning of registration entries as an example here https://github.com/spiffe/spire/blob/9c11127705aa706d0da9783c08579165b7b05d68/pkg/server/datastore/sqlstore/sqlstore.go#L4087-L4091